core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -211,7 +211,23 @@ static Expression fromInternal(Expression e, Class<?> targetType) {
       List<Expression> expressions) {
     final List<Expression> list = new ArrayList<>();
     for (int i = 0; i < expressions.size(); i++) {
-      list.add(fromInternal(expressions.get(i), targetTypes[i]));
+      if (targetTypes[i].isArray()) {
+        // only the last type could be variable length argument. All left expressions
```

Comment:
Only the last can be a vararg but there can be other array types in between, does this pose a problem? Maybe add some unit tests with such cases?

Reply:
Actually, this is the one comment that I care most. The rest are minor and I could treat them myself I think.

Reply:
Your concern is valid. I added a check to check if it's checking the last type to decide if move forward for varargs logic. 

Note that only the last type could be varargs. Complier guarantees it. Array types have values as arrays. So except for varargs, other types (including array types) are still 1:1 mapped to values.

Reply:
Have added more tests for this case.